File: sentimental-credit/credit.py
from math import floor
import re


def get_size(userInput):
    if userInput > 0:
        string_rep = str(userInput)
        return len(string_rep)
    return 0

# Check that the card is infact a valid card


def checksum(usr_input, inputSize):
    # x2Sum will be used to aggregate the digits multiplied by 2
    x2Sum = 0
    secondToLastIndex = inputSize - 2
    for i in range(secondToLastIndex, -1, -2):
        # if the value of index i is >= 10 split it into 2 and then add otherwise add it as is to x2Sum
        if usr_input[i] * 2 < 10:
            x2Sum += usr_input[i] * 2
        else:
            rem = int(floor(usr_input[i] * 2) % 10)
            quo = int(floor(usr_input[i] * 2) / 10)
            x2Sum += rem + quo
    sum = 0
    for i in range(inputSize - 1, -1, -2):
        sum += usr_input[i]
    total = x2Sum + sum
    # if total is a multiple of 10
    if int(floor(total % 10 == 0)):
        return True
    return False

# Check if the card is an American Exchange card


def IsAE(usr_input,  inputSize):
    if (inputSize != 15):
        return False
    # Yes at this point usr_input is valid so you can check at aindex 0 and 1
    if ((usr_input[0] == 3 and usr_input[1] == 4) or (usr_input[0] == 3 and usr_input[1] == 7)):
        return True
    return False

# / Check if the card is a VisaCard


def IsVC(usr_input,  inputSize):
    if (inputSize != 16 and inputSize != 13):
        return False
    if (usr_input[0] == 4):
        return True
    return False


# Check if the card is a MasterCard
def IsMC(usr_input,  inputSize):
    if (inputSize != 16):
        return False
    if (
        (usr_input[0] == 5 and usr_input[1] == 1) or
        (usr_input[0] == 5 and usr_input[1] == 2) or
        (usr_input[0] == 5 and usr_input[1] == 3) or
        (usr_input[0] == 5 and usr_input[1] == 4) or
            (usr_input[0] == 5 and usr_input[1] == 5)):
        return True
    return False


def main():
    raw_input = input("Input your card number: ")
    valid = re.fullmatch("^[0-9]*", raw_input)
    if (not valid):
        print("INVALID\n")
        return 0
    card_num = int(valid.string)
    count = get_size(card_num)
    if (count < 13 or count > 16 or count == 14):
        print("INVALID\n")
        return 0
    split_input = [0] * count
    # get a power of 10 which we will use to split the usr_input into integer singletons.
    # i.e a number with a size of 13 needs 10^12 to get the 1st number.
    count_2 = pow(10, count - 1)

    # tracks the index to store the parsed number in the split_input array.
    index = 0

    # now put these singeltons in an array
    while (count_2 >= 1):
        temp = card_num / count_2
        rem = card_num % count_2
        split_input[index] = int(floor(temp))
        card_num = rem
        index += 1

        # reduce count_2 by a factor of /10
        count_2 = int(floor(count_2 / 10))

    # if the card is at least a valid card then procede to checking if it is one of them after which you print out the corresponding name.
    if (checksum(split_input, count)):
        if (IsAE(split_input, count)):
            print("AMEX\n")
        elif (IsMC(split_input, count)):
            print("MASTERCARD\n")
        elif (IsVC(split_input, count)):
            print("VISA\n")
        else:
            print("INVALID\n")
    else:
        print("INVALID\n")

    return 0


if __name__ == '__main__':
    main()

File: credit/credit.c
#include <cs50.h>
#include <math.h>
#include <string.h>
#include <stdio.h>

// check that the user input is unsigned
bool input_is_unsigned(long userInput)
{
    if (userInput > 0)
    {
        return true;
    }
    return false;
}

// Gets the length of the user input
int get_size(long userInput)
{
    if (input_is_unsigned(userInput))
    {
        char string_rep[100];
        sprintf(string_rep, "%ld", userInput);
        return strlen(string_rep);
    }
    return 0;
}


// Check that the card is infact a valid card
bool checksum(int input[], int inputSize)
{

    // x2Sum will be used to aggregate the digits multiplied by 2
    int x2Sum = 0;

    int secondToLastIndex = inputSize - 2;
    for (int i = secondToLastIndex; i >= 0; i -= 2)
    {
        // if the value of index i is >= 10 split it into 2 and then add otherwise add it as is to x2Sum
        if (input[i] * 2 < 10)
        {
            x2Sum += input[i] * 2;
        }
        else
        {
            int rem = (input[i] * 2) % 10;
            int quo = (input[i] * 2) / 10;
            x2Sum += rem + quo;
        }
    }


    int sum = 0;
    for (int i = inputSize - 1; i >= 0; i -= 2)
    {
        sum += input[i];
    }

    int total = x2Sum + sum;

    // if total is a multiple of 10
    if (total % 10 == 0)
    {
        return true;
    }
    return false;
}

/// Check if the card is an American Exchange card
bool IsAE(int input[], int inputSize)
{
    if (inputSize != 15)
    {
        return false;
    }
    // Yes at this point input is valid so you can check at aindex 0 and 1
    if ((input[0] == 3 && input[1] == 4) || (input[0] == 3 && input[1] == 7))
    {
        return true;
    }
    return false;
}

/// Check if the card is a VisaCard
bool IsVC(int input[], int inputSize)
{
    if (inputSize != 16 && inputSize != 13)
    {
        return false;
    }

    if (input[0] == 4)
    {
        return true;
    }
    return false;
}

// Check if the card is a MasterCard
bool IsMC(int input[], int inputSize)
{
    if (inputSize != 16)
    {
        return false;
    }

    if (
        (input[0] == 5 && input[1] == 1) ||
        (input[0] == 5 && input[1] == 2) ||
        (input[0] == 5 && input[1] == 3) ||
        (input[0] == 5 && input[1] == 4) ||
        (input[0] == 5 && input[1] == 5))
    {
        return true;
    }
    return false;
}

int main(void)
{

    long input = get_long("Input your card number: ");
    int count = get_size(input);

    if (count < 13 || count > 16 || count == 14)
    {
        printf("INVALID\n");
        return 0;
    }
    //
    int split_input[count];

    // get a power of 10 which we will use to split the input into integer singletons.
    // i.e a number with a size of 13 needs 10^12 to get the 1st number.
    long count_2 = pow(10, count - 1);

    // tracks the index to store the parsed number in the split_input array.
    int index = 0;

    // now put these singeltons in an array
    while (count_2 >= 1)
    {
        long temp = input / count_2;
        long rem = input % count_2;

        split_input[index] = temp;
        input = rem;
        index += 1;

        // reduce count_2 by a factor of /10
        count_2 = count_2 / 10;
    }

    // if the card is at least a valid card then procede to checking if it is one of them after which you print out the corresponding name.
    if (checksum(split_input, count))
    {
        if (IsAE(split_input, count))
        {
            printf("AMEX\n");
        }

        else if (IsMC(split_input, count))
        {
            printf("MASTERCARD\n");
        }
        else if (IsVC(split_input, count))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    else
    {
        printf("INVALID\n");
    }
    return 0;
}